src/foraging_gui/TransferToNWB.py

```diff
@@ -340,83 +340,100 @@ class obj:
+        'laser_condition_probability' : LaserConditionProC,
+        'laser_start' : LaserStartC,
+        'laser_start_offset' : LaserStartOffsetC,
+        'laser_end' : LaserEndC,
+        'laser_end_offset' : LaserEndOffsetC,
+        'laser_protocol' : LaserProtocolC,
+        'laser_frequency' : LaserFrequencyC,
+        'laser_rampingdown' : LaserRampingDownC,
+        'laser_pulse_duration' : LaserPulseDurC,
+
+        'session_wide_control' : _get_field(obj, 'TP_SessionWideControl', index=i, default='None'),
+        'fraction_of_session' : float(_get_field(obj, 'TP_FractionOfSession', index=i, default=np.nan)),
+        'session_start_with' : _get_field(obj, 'TP_SessionStartWith', index=i, default='None'),
+        'session_alternation' : _get_field(obj, 'TP_SessionAlternating', index=i, default='None'),
+        'minimum_opto_interval' : float(_get_field(obj, 'TP_MinOptoInterval', index=i, default=0)),
+
+        # add all auto training parameters (eventually should be in session.json)
+        'auto_train_engaged' : _get_field(obj, 'TP_auto_train_engaged', index=i, default='None'),
+        'auto_train_curriculum_name' : _get_field(obj, 'TP_auto_train_curriculum_name', index=i, default='None'),
+        'auto_train_curriculum_version' : _get_field(obj, 'TP_auto_train_curriculum_version', index=i, default='None'),
+        'auto_train_curriculum_schema_version' : _get_field(obj, 'TP_auto_train_curriculum_schema_version', index=i,
+                                                          default='None'),
+        'auto_train_stage' : _get_field(obj, 'TP_auto_train_stage', index=i, default='None'),
+        'auto_train_stage_overridden' : _get_field(obj, 'TP_auto_train_stage_overridden', index=i, default=np.nan),
+
+        # reward size
+        'reward_size_left' : float(_get_field(obj, 'TP_LeftValue_volume', index=i)),
+        'reward_size_right' : float(_get_field(obj, 'TP_RightValue_volume', index=i))
+        }
+
+        # populate lick spouts with correct format desppending if using newscale or aind stage
+        stage_positions = getattr(obj, 'B_StagePositions', [])
+        if i < len(stage_positions):    # index is valid for stage position lengths
+            trial_kwargs['lickspout_position_x'] = stage_positions[i]['x']
+            trial_kwargs['lickspout_position_z'] = stage_positions[i]['z']
+            if stage_positions[i].keys() == ['x', 'y', 'z']:    # newscale stage
+                trial_kwargs['lickspout_position_y'] = stage_positions[i]['y']
+            else:       # aind stage
+                trial_kwargs['lickspout_position_y1'] = stage_positions[i]['y2']
+                trial_kwargs['lickspout_position_y2'] = stage_positions[i]['y2']
+        else:   # if i not valid index, populate values with nan
+            trial_kwargs['lickspout_position_x'] = np.nan
+            trial_kwargs['lickspout_position_y'] = np.nan
+            trial_kwargs['lickspout_position_z'] = np.nan
+        
```

Comment:
@alexpiet changed so a dictionary of relevant info is created and passed into add_trial. This way, only correct stage axis are added to nwb. This logic block first checks if i is a valid index in B_StagePositions. If the obj does not have a B_StagePositions attribute, the default is empty list so logic block skips to else because i is not a valid index. Otherwise, the lickspout_position keys are created based on the keys in B_StagePositions. The default will always be
 `trial_kwargs['lickspout_position_x'] = np.nan
  trial_kwargs['lickspout_position_y'] = np.nan
  trial_kwargs['lickspout_position_z'] = np.nan`
if B_StagePositions isn't an attribute in obj or the dictionary keys are incorrect. Let me know if that is not a good default or if this is not the right approach

Reply:
makes sense

Reply:
After testing though I'm getting value errors for lickspout_position_y1, lickspout_position_y2 and missing lickspout_position_y. Is this something intrensic to the nwb.add_trial function or is it a schema on our side? I'm not super familiar with nwb files

Reply:
You just need to modify this line: https://github.com/AllenNeuralDynamics/dynamic-foraging-task/blob/cb95f267f15042ee4efe4ca2bbcae0e09ecffaef/src/foraging_gui/TransferToNWB.py#L275

Which is defining what columns should be added to the trials table, so then when you call "add_trial" its complaining because there is a mis-match.

Reply:
ahh great thank you

Reply:
Okay now tested and working for json with B_StagePositions with 'x', 'y', 'z' keys and 'x', 'y1', 'y2', 'z' keys. Also working in case that B_StagePositions doesn't exist or has incorrect keys